[NO QA] Add PR review checklists for contributors & C+ (#7565)
Expensify/App: PR_REVIEW_CHECKLIST.md

# PR review checklist

> This document lists specific checks that should be done when reviewing a PR. This means these items should be checked by:

Reviewer: ```suggestion
This document lists specific guidelines for all PR reviews and aims to clarify items in the PR checklist.
```
Since this document is not a checklist I think this change would make more sense. If these guidelines apply to all reviewers then we should just say that and remove the numbered list below.

Author: Nice, since marc gave similar feedback I'll remove the itemized list below :)

# PR review checklist

This document lists specific checks that should be done when reviewing a PR. This means these items should be checked by:
1. Every contributor submitting a Pull Request
2. Every C+ (Contributor Plus) team member reviewing a PR

> 3. Every Expensify employee submitting / reviewing a PR

Author: I see what you mean, but I think it's nice to be explicit here so that it's **clear** to contributors & C+ that they need to know & be following this checklist all the time.

> This is a living document, so we will be updating this from time to time. Feel free to suggest additions or changes in the #expensify-open-source Slack channel.

Author: I thought it would be useful to at least clarify that the process will be changing in the next few weeks / months as we get everything nailed down. I agree we don't necessarily want to advertise giving suggestions, we can do that in C+ channel if we want their ideas

# Check PR template

> 1. Make sure the PR template includes screenshots or videos of the changes being tested on **all platforms**. If there are one or more platforms missing, follow up with the contributor to determine why - we shouldn't be merging changes that aren't tested everywhere.

Reviewer 2: Is this kind of redundant? Shouldn't we only put things into this file that are not covered in the PR template? If we change something about this later we have to update it in two places.

Author: Good idea, it's mentioned in the template AND in Contributing.md so i'll remove it from here

    - > **Rule of thumb**: Over-explain what the tester needs to do. Do not rely on the tester having existing knowledge of the app.
      > 
      > ### Example of bad steps:
      >
	  >	1. Go to this page
	  >	2. Click the button
      >
      > ### Example of good steps:
      >
      > 1. Log in with an account that has a workspace, or follow `[these steps](link-to-instructions-for setting-up-workspace)` to make one.
      > 2. Click the settings button on the top left (the one with your avatar)
      > 3. Navigate to “Your Workspace” > Reimburse Receipts > Connect bank account
      > 4. Verify that the bank account view opens without errors
      >
    - Take note of the distinction between testing _locally_ and _on staging_. Note: The staging site references the production version of the API.
4. Make sure the testing steps also cover possible regressions.
    1. Cases where we definitely need new regression tests:
        1. Building a new page
        2. Building a new feature
        3. Building a new component?
    2. Cases where we might need new regression tests:

>         1. Bug was easy to find

Reviewer 2: I still think this section is confusing. How does a contributor know they aren't suggesting regressions we already have? I think this should be between internal engineers and Applause for now. Just my 2 cents though.

Reviewer 2: Should we at least move it into the next section? We are talking about testing here and also in the `Testing the changes` section. Maybe we can find a better way to organize this information?

Author: Regression discussion on hold, asking in #applause-internal if Applause may be able to take of this for us & for contributors

Author: Removing for now, regression testing discussion pushed to future haha

5. Make sure that the test includes both **success** and **fail** scenarios.
    - Ex: if the issue was to prevent the user from saving a field when it is empty, besides testing that we should also test that they can save the field _if it is not empty_.

# Testing the changes

1. Make sure there are no console errors related to the PR. If you find console errors in a branch you’re reviewing, check if they also exist in `main`.
    1. If they do, proceed as normal (report them in #expensify-open-source if you think they need to be fixed).
    2. If the errors do not exist in `main`, report it to the PR author - they need to be fixed before the PR can be merged.
2. Test the changes on **all platforms**. Follow this guide (needs link) to set up each platform.
    - If you’re unable to boot a platform for any reason, ask for help in the #expensify-open-source Slack channel. Your issue might already have been addressed there, so be sure to search first.

> 3. Should we add regression testing for code introduced in this PR?

Author: Some discussion about regressions [here](https://docs.google.com/document/d/1G-WFzlsc0wqxgbhs8ejP0Ry-vechEIrlKj1FIdaTf20/edit?disco=AAAAUdQgVGw) - between @marcaaron and @Luke9389

Reviewer 3: Still discussing on the doc. At a high level, I think we want to raise awareness of side effects and try to encourage the active prevention of regressions. How exactly that happens has yet to be decided it seems

Reviewer 2: My feelings are the same. If we can't be clear about what we are asking people to do - I'm not sure we should ask them to do it at all. That introduces ambiguity and seems like the opposite of what we are trying to achieve with these guidelines.

Author: Killing all regression test mentions for now - will bring up this convo again later in Slack :)

    - If the PR introduces **new** pages, features, etc. then we **should** have new regression tests written. Please note this in your PR so that Applause can make sure it gets added.

>     - If the PR fixes a bug, Applause will check to make sure the existing regression testing steps already cover the workflow fixed by the bug.

Reviewer 2: This possibly seems like something we should just ask Applause to do. It seems like a weird instruction to ask someone to:

- Suggest that a test be added as a regression
- Applause checks if it exists 
- Applause adds it as a regression if not

Why is Applause not identifying the opportunity to just add the regression test? Is this defined in the scope of the work they are contracted to do for us, or not? If not, and we really need contributors to suggest areas to test for regressions we should provide examples of WHAT that task looks like and WHO is responsible to do this.

>Please note this in your PR so that Applause can make sure it gets added.

What does this mean? How does someone "note" this and who are we talking to?

Author: > Is this defined in the scope of the work they are contracted to do for us, or not?

This seems like the main question - I agree if they can / are willing to do it themselves, that would definitely be best. Then the other questions don't really need to be answered 😅

Author: Asking in #applause [here](https://expensify.slack.com/archives/CAMPKJQ94/p1644235856338909)

Author: Hmm @tgolen made a good point in Slack ([here](https://expensify.slack.com/archives/CAMPKJQ94/p1644251179988849?thread_ts=1644235856.338909&cid=CAMPKJQ94)):

> We are the experts on knowing what's valuable to test and what isn't. If we just gave them open permission to create tests, it could result in a bunch of bloat that just increases the price of our contract. It would be in their best interest to create as many tests as they could so they could get paid more.

# Reviewing the code

## Good code patterns to require

1. Check that functions have comments when appropriate.
    - If the function has params or returns something, use JSDocs syntax to describe them.
        - Indicate the param name(s), data type passed and / or returned, and purpose if not immediately obvious.
    - Obvious functions (with no params / return value) should not have comments.

>     - **In short: _Add comments & docs often, only when useful._**

Reviewer 2: Remove the "often". Comments should not be encouraged in any particular frequency.

2. All copy / text shown to users in the product should be stored in the `src/languages/*` files for localization.
3. Platform-specific files should follow the proper naming convention (see [Platform-Specific File Extensions](https://github.com/expensify/app#platform-specific-file-extensions))
    - Example: `MyComponent/index.android.js`

## Code patterns to check for improvement

> 1. Is there appropriate error handling?

Reviewer 2: Huge can of worms. What are we actually asking here?

Author: True, we need real direction here - what is "appropriate error handling" in our App? I propose we remove this but take the discussion to Slack (probs #contributor-management). I think it's important we bring this back to the checklist eventually, because there's lots of different ideas of how to do error handling (and people can easily be lazy and skip it)

> 2. Reuse...

Reviewer 2: ?

Author: Oops this was a placeholder, removing

> 3. Are there any changes that could've been applied to one component that were applied to every usage of the component instead?

Reviewer 2: I don't understand this instruction and it feels unnecessarily specific. Are we asking if there is some opportunity to remove redundant code specific to a React component? Why not just generally ask people to address opportunities to remove redundant code? This is basically just the DRY principle.

Author: > Are we asking if there is some opportunity to remove redundant code specific to a React component?

Yeah basically -> Checking if there's a way to refactor the component (or make a new one) to DRY up code around where the component is being used.

I think you're right it's basically DRY & a repeat of the "Considerations around code reuse" topic, so I think this can be removed.

Think it's useful to mention the DRY principle in this checklist or is it pretty self explanatory?

Author: Removing in favor of only including clear points

>     - Example: https://github.com/Expensify/App/pull/3129#pullrequestreview-668271068
> 4. Ensure fixes for just android or ios aren't put in a `index.native.js`. Instead they should be in a specific `index.ios.js` or `index.android.js` file.
>     - Put another way, ensure platform-specific solutions are not applied cross-platform.

Reviewer 2: ```suggestion
    - Example: https://github.com/Expensify/App/pull/3129#pullrequestreview-668271068
4. Platform-specific solutions should not be applied to more platforms than necessary
    - e.g. fixes for Android or iOS should go in specific `index.ios.js` or `index.android.js` files and not `index.native.js` files
```

5. Look for usages of react fragment wrappings and confirm they are necessary.

>     - Also look for any unnecessary `View`s and think if they should be replaced with React fragments.

Author: Some discussion in the doc between @marcaaron and @roryabraham regarding views / fragments

Reviewer 2: I don't think this is valuable to mention. It's on the extreme lower end of issues that threaten the health of this codebase.

Author: I think I agree - if it becomes an issue we can add it later, better to keep this list short & incredibly useful for now, and add things later if we want

> 6. Were any new styles added? Can we reuse existing styles instead? Can we generalize the new style?

Reviewer 2: Let's link to the `STYLING.md` doc here rather than ask all these questions just say

```
6. Were any new styles added? Are we following the guidelines in [`STYLING.md`](..linktoit)?
```

Author: Agreed, I like this

## Code patterns to avoid

> 1. Make sure any `setTimeout`s are absolutely necessary, not just a workaround for a situation that isn’t fully understood.

Reviewer 2: This reads a bit like "if you understand why a setTimeout is needed then they are OK to use"

Reviewer 2: FWIW the most common example where we see this is to solve a race condition.

Author: > This reads a bit like "if you understand why a setTimeout is needed then they are OK to use"

That's why I put "absolutely necessary" 🤔  But isn't a `setTimeout` ok to keep if it's "absolutely necessary" and the author & reviewer understand why?

Reviewer: ```suggestion
1. Make sure any `setTimeout`s are absolutely necessary, not just a workaround for a situation that isn’t fully understood (e.g., a race condition).
```
How about that?

Author: Nicee 👍 making the change!

2. Ensure Onyx is used appropriately. ([docs](https://github.com/expensify/react-native-onyx#merging-data))

> 3. Don't mention the `isSmallScreenWidth` prop type explicitly in any component, use `...windowDimensionsPropTypes` instead.

Reviewer 2: I still don't understand this and it doesn't seem valuable to mention. Doesn't even make my list of nitpicking concerns personally.

Author: Personally, I would really prefer if we nail down what to do in this situation. If a component only uses 1 prop passed by an HOC (`withWindowDimensions`, for example), do we:
1. Document only the prop that we care about & use in that component; or
    - This could be not ideal b/c we would most likely repeat the same PropType definition in many different components -> However, that could be avoided by exporting multiple individual prop type definitions from the reusable PropTypes file 
2. Spread the corresponding reusable propTypes file (`windowDimensionsPropTypes` in our example)

Author: Removing for now, if we want to add it back in later we can

## Considerations around code reuse

> While trying to keep our code DRY (don't repeat yourself), you may find you're walking a thin line between refactoring & reusing a component vs. creating a brand new component. Here are some guidelines we use - they're not hard and fast rules, so please feel free to post in #expensify-open-source to gather peer opinions in order to make the best decisions.

Reviewer 2: NAB, idioms like "walk a thin line" add flavor but are less direct than just telling people straight up what DRY means, how to identify opportunities for it, etc

Author: Agreed, I thought we agreed that this topic (specialization vs reuse) wasn't 100% clear, meaning it won't be easy to give super direct instruction. Would this be a better intro:

> While trying to keep our code DRY (don't repeat yourself), you may consider refactoring & reusing a component vs. creating a brand new specialized component. Here are some guidelines we use - they're not hard and fast rules, so please feel free to post in #expensify-open-source to gather peer opinions in order to make the best decisions.

> 1. Specialization

Author: @Luke9389 I got this from your section "When to ask for help" - I decided to focus on code reuse ideas (specialize vs refactor) but feel free to suggest revisions :D

Reviewer 3: I'd love for @marcaaron to check this out bc I was sort of adapting some of his comments from a few threads. I'm totally open to changes here. In general, I like the idea of an actual "When to Ask For Help" section that covers the stuff that doesn't neatly fit into checkboxes.

Reviewer 2: This is great!

Author: @Luke9389 Do you think the "When to ask for help" can be super generic and can be used to help us improve the doc? Like generically "when to ask for help on a specific checklist item" or "when to ask for help on a specific PR b/c there's some considerations that aren't outlined in the doc yet"?

Reviewer 3: I don't think I understand your question fully, but I'd prefer the section to cover stuff that's not on a checkbox. So more the latter.

    - When one component is a special case of another, we should opt for the technique of composition. Here are the [React docs](https://reactjs.org/docs/composition-vs-inheritance.html#specialization) on Specialization. The idea is that it’s better to establish a pattern of creating increasingly specific components, instead of adding dozens of use cases to a single component.

>     - You might consider this when a component is being reused because the new use case is “close enough” to the original. Rather than adding functionality to that component with subtle additions, it may be better to create a new, more sprecialized version of that component.

Reviewer 2: Amazing section. Love all of this.

>     - You might consider this when a component is being reused because the new use case is “close enough” to the original. Rather than adding functionality to that component with subtle additions, it may be better to create a new, more sprecialized version of that component.

Reviewer 2: `specialized` has a typo

1. Refactors
    - During a code review it becomes apparent that a refactor is needed. In this case, we recommend following these steps:
        1. Identify the refactor and confirm with others that it’s needed.
        2. Do that refactor first (as a separate job, if it qualifies), merge it, test for regressions.
        3. Proceed with the original issue.

# Submit / approve the PR

> While the steps above can be tedious to follow the first time you submit or review a PR, over time they will become more natural. The goal of these steps is to make sure we create the best product, with the best engineers writing the best code.

Reviewer 2: Is it necessary to suggest that any part of this is tedious, unnatural, or difficult to follow? 

It's the job and this is what is required. If it is hard to follow then we are not doing a good job at communicating what needs to be done and should improve this document.

Author: "necessary"? Definitely not :D Maybe I like to explain things via [shit sandwhich](https://en.wiktionary.org/wiki/shit_sandwich) too much haha

Author: > If it is hard to follow then we are not doing a good job at communicating what needs to be done and should improve this document.

I don't think I was implying that this checklist is hard to follow, just that it can be time consuming the first time it's looked again. Again, I don't think it's necessary so I'm still happy to chop it

Reviewer 2: Yeah, let's just assume that we are not asking for something that is even remotely painful (everyone will have their own opinion regardless). The idea behind the checklist is that it will make contributor jobs EASIER and not tedious.

Author: Good idea, I'll just remove this whole final section since it's definitely not important :D
